### PR TITLE
Reserved Pages convenience improvements

### PR DESCRIPTION
**Pull request by toly-kournik**
- Initialize ReservedPagesCLient only once instead of initializing every client with ReservedPages;
- ReservedPagesClient to implemet IReservedPages;
- Make res pages offset implicit making each RP client to use relative page id instead of explicitly adding its resPagesOffset()

### CODE REVIEW

File: bftengine/src/bftengine/BFTEngine.cpp [hunk 1]
```diff
@@ -22,10 +22,13 @@
 #include "PreProcessor.hpp"
 #include "MsgReceiver.hpp"
 #include "RequestHandler.h"
+#include "ReservedPagesClient.hpp"
 #include <condition_variable>
 #include <memory>
 #include <mutex>
 
+bftEngine::IReservedPages *bftEngine::ReservedPagesClientBase::res_pages_ = nullptr;
```
Comment (yontyon) [inline bftengine/src/bftengine/BFTEngine.cpp:30]: Make sure it doesn't create problems with the product. A few weeks ago we had issues with global static variables and eventually, we forced to define them on the stack

File: bftengine/src/bftengine/ControlStateManager.cpp [hunk 2]
```diff
@@ -89,7 +85,7 @@ void ControlStateManager::clearCheckpointToStopAt() {
   std::ostringstream outStream;
   concord::serialize::Serializable::serialize(outStream, tmpPage);
   auto data = outStream.str();
-  reserved_pages_->saveReservedPage(resPageOffset(), data.size(), data.data());
+  saveReservedPage(0, data.size(), data.data());
```
Comment (yontyon) [inline bftengine/src/bftengine/ControlStateManager.cpp:88]: Why do we have a hardcoded 0?

**Review by yontyon — APPROVED** [review]
LGTM!
Can you have a dummy MR with this branch before merging it? 
I just to be sure that we don't run into static variables problem in the product again

File: bftengine/src/bftengine/BFTEngine.cpp (hunk 1 above)
Comment (toly-kournik) [reply bftengine/src/bftengine/BFTEngine.cpp:30]: it's a class static variable

File: bftengine/src/bftengine/ControlStateManager.cpp (hunk 2 above)
Comment (toly-kournik) [reply bftengine/src/bftengine/ControlStateManager.cpp:88]: It's a relative offset, meaning page 0

**Pull request merged**